runtime/ints.c

```diff
@@ -280,6 +280,157 @@ CAMLprim value caml_int32_shift_right(value v1, value v2)
 CAMLprim value caml_int32_shift_right_unsigned(value v1, value v2)
 { return caml_copy_int32((uint32_t)Int32_val(v1) >> Int_val(v2)); }
 
+#if defined(__GNUC__)
+#if ARCH_INT32_TYPE == long
+#define int32_clz __builtin_clzl
+#define int32_popcnt __builtin_popcountl
+#else /* ARCH_INT32_TYPE == long */
+#define int32_clz __builtin_clz
+#define int32_popcnt __builtin_popcount
+#endif /* ARCH_INT32_TYPE == long */
+
+#define int64_clz __builtin_clzll
+#define int64_popcnt __builtin_popcountll
+
+#else /* defined(__GNUC__) */
+ /* _MSC_ does not have builtin for popcnt, and
+  *  has undefined behavior for clz builtin when
+  *  lzcnt instruction is not supported.
+  *  Use naive version of clz and popcnt from Hacker's Delight. */
```

Comment:
I thought you could use the `__popcnt` intrinsics, see https://docs.microsoft.com/en-us/cpp/intrinsics/popcnt16-popcnt-popcnt64.

For `clz`, there is https://docs.microsoft.com/en-us/cpp/intrinsics/bitscanforward-bitscanforward64

Reply:
It looks like popcnt intrinsic is not available on all architectures, according to this list:
https://docs.microsoft.com/en-us/cpp/intrinsics/intrinsics-available-on-all-architectures?view=vs-2017. For clz, bit scan reverse can be used, although it is not available on some CPUs. A special case would be needed anyway for i386 and a check of cpuid for support on amd64.  
It is only used for a slow path (not native code) so it seemed cleaner to have naive code.